packages/fileeditor/src/toc/markdown.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { DocumentRegistry, IDocumentWidget } from '@jupyterlab/docregistry';
+import {
+  TableOfContents,
+  TableOfContentsModel,
+  ToCUtils
+} from '@jupyterlab/toc';
+import { Widget } from '@lumino/widgets';
+import { FileEditor } from '../widget';
+import { EditorToCModelFactory, IEditorHeading } from './factory';
+
+/**
+ * Table of content model for Markdown files.
+ */
+export class MarkdownToCModel extends TableOfContentsModel<
+  IEditorHeading,
+  IDocumentWidget<FileEditor, DocumentRegistry.IModel>
+> {
+  /**
+   * Type of document supported by the model.
+   *
+   * #### Notes
+   * A `data-document-type` attribute with this value will be set
+   * on the tree view `.jp-TableOfContents-content[data-document-type="..."]`
+   */
+  get documentType(): string {
+    return 'markdown';
+  }
+
+  /**
+   * Produce the headings for a document.
+   *
+   * @returns The list of new headings or `null` if nothing needs to be updated.
+   */
+  protected getHeadings(): Promise<IEditorHeading[] | null> {
+    if (!this.isActive) {
+      return Promise.resolve(null);
+    }
+
+    const content = this.widget.content.model.value.text;
+
+    const headings = ToCUtils.Markdown.getHeadings(content, {
+      ...this.configuration,
+      // Force removing numbering as they cannot be displayed
+      // in the document
+      numberHeaders: false
+    });
+    return Promise.resolve(headings);
+  }
+}
+
+/**
+ * Table of content model factory for Markdown files.
+ */
+export class MarkdownToCFactory extends EditorToCModelFactory {
```

Comment:
Since these are exposed as part of the public API, wondering whether we should keep the full name for clarity. For example:

```suggestion
export class MarkdownTableOfContentsFactory extends EditorTableOfContentsModelFactory {
```

I suppose "ToC" is used here to keep the name shorter?

Reply:
Indeed that was by laziness :wink: I'll change the names to be more explicit